src/libraries/System.Private.CoreLib/src/System/IO/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -0,0 +1,406 @@
+        }
+
+        protected override void OnInitFromHandle(SafeFileHandle handle)
+        {
+            // This is necessary for async IO using IO Completion ports via our
+            // managed Threadpool API's.  This calls the OS's
+            // BindIoCompletionCallback method, and passes in a stub for the
+            // LPOVERLAPPED_COMPLETION_ROUTINE.  This stub looks at the Overlapped
+            // struct for this request and gets a delegate to a managed callback
+            // from there, which it then calls on a threadpool thread.  (We allocate
+            // our native OVERLAPPED structs 2 pointers too large and store EE
+            // state & a handle to a delegate there.)
+            //
+            // If, however, we've already bound this file handle to our completion port,
+            // don't try to bind it again because it will fail.  A handle can only be
+            // bound to a single completion port at a time.
+            if (!(handle.IsAsync ?? false))
+            {
+                try
+                {
+                    handle.ThreadPoolBinding = ThreadPoolBoundHandle.BindHandle(handle);
+                }
+                catch (Exception ex)
+                {
+                    // If you passed in a synchronous handle and told us to use
+                    // it asynchronously, throw here.
+                    throw new ArgumentException(SR.Arg_HandleNotAsync, nameof(handle), ex);
+                }
+            }
+        }
+
+        protected override void OnInit()
+        {
+            // This is necessary for async IO using IO Completion ports via our
+            // managed Threadpool API's.  This (theoretically) calls the OS's
+            // BindIoCompletionCallback method, and passes in a stub for the
+            // LPOVERLAPPED_COMPLETION_ROUTINE.  This stub looks at the Overlapped
+            // struct for this request and gets a delegate to a managed callback
+            // from there, which it then calls on a threadpool thread.  (We allocate
+            // our native OVERLAPPED structs 2 pointers too large and store EE state
+            // & GC handles there, one to an IAsyncResult, the other to a delegate.)
+            try
+            {
+                _fileHandle.ThreadPoolBinding = ThreadPoolBoundHandle.BindHandle(_fileHandle);
+            }
+            catch (ArgumentException ex)
+            {
+                throw new IOException(SR.IO_BindHandleFailed, ex);
+            }
+            finally
+            {
+                if (_fileHandle.ThreadPoolBinding == null)
+                {
+                    // We should close the handle so that the handle is not open until SafeFileHandle GC
+                    Debug.Assert(!_exposedHandle, "Are we closing handle that we exposed/not own, how?");
+                    _fileHandle.Dispose();
+                }
+            }
+        }
+
+        // called by BufferedStream
```

Comment:
One idea I had was to replace the current logic (reuse `PreAllocatedOverlapped` when buffering is enabled) with caching `PreAllocatedOverlapped` and reusing it if the user (or some strategy) provided the same buffer as before. It could help in a scenario where the user is just reading the entire file:

```cs
public async Task<long> ReadAsync_NoBuffering(long fileSize, int userBufferSize, FileOptions options)
{
    Memory<byte> userBuffer = new Memory<byte>(_userBuffers[userBufferSize]);
    long bytesRead = 0;
    using (FileStream fileStream = new FileStream(_sourceFilePaths[fileSize], FileMode.Open, FileAccess.Read, FileShare.Read, 1, options))
    {
        while (bytesRead < fileSize)
        {
            // buffering is disabled, but we could detect that user is passing the same `userBuffer` all the time
            bytesRead += await fileStream.ReadAsync(userBuffer, CancellationToken.None);
        }
    }

    return bytesRead;
}
```

We could make it a weak reference to ensure we don't leak the memory.

@stephentoub what do you think?